Add REPL mode for cli.

tirkarthi opened the pull request:
## Summary

click commands can be wrapped in a repl sub-command where we can start a repl to access all other commands. This should persist session and also supports auto-complete along with storing history.

## Changes

* Add [click-repl](https://github.com/click-contrib/click-repl) as requirement
* Add repl command to start repl

https://user-images.githubusercontent.com/3972343/160831507-2d24edb1-7044-4476-af0b-e45e92dc0a9c.mp4

```
prefect repl

Welcome to Prefect REPL
version : 1.1.0+10.g7d958a5e7.dirty
    
> get projects
NAME    FLOW COUNT    AGE             DESCRIPTION
Sample  1             45 minutes ago
> get flows -p Sample
NAME         VERSION    AGE             ID                                    PROJECT NAME
Hello world  1          45 minutes ago  1aded56d-c6c2-4939-a247-cdc59d23de47  Sample
> run --id 1aded56d-c6c2-4939-a247-cdc59d23de47
Looking up flow metadata... Done
Creating run for flow 'Hello world'... Done
└── Name: opalescent-degu
└── UUID: 57c7dc27-edf5-44de-8b71-751073a3efe1
└── Labels: ['laptop']
└── Parameters: {}
└── Context: {}
└── URL: http://localhost:8080/default/flow-run/57c7dc27-edf5-44de-8b71-751073a3efe1
> get logs --id 57c7dc27-edf5-44de-8b71-751073a3efe1
TIMESTAMP                         LEVEL    MESSAGE
2022-03-30T12:04:09.092327+00:00  INFO     Submitted for execution: PID: 33425
2022-03-30T12:04:09.559358+00:00  INFO     Beginning Flow run for 'Hello world'
2022-03-30T12:04:09.687275+00:00  INFO     Task 'hello': Starting task run...
2022-03-30T12:04:09.812571+00:00  INFO     Task 'hello': Finished task run for task with final state: 'Success'
2022-03-30T12:04:09.83998+00:00   INFO     Task 'hello': Starting task run...
2022-03-30T12:04:09.917872+00:00  INFO     Task 'hello': Finished task run for task with final state: 'Success'
2022-03-30T12:04:09.944946+00:00  INFO     Task 'world': Starting task run...
2022-03-30T12:04:10.017424+00:00  INFO     Task 'world': Finished task run for task with final state: 'Success'
2022-03-30T12:04:10.033041+00:00  INFO     Flow run SUCCESS: all reference tasks succeeded
```

## Importance

This will improve the usability of the cli.

## Checklist

This PR:

- [ ] adds new tests (if appropriate)
- [ ] adds a change file in the `changes/` directory (if appropriate)
- [ ] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

madkinsz commented on src/prefect/cli/__init__.py:
```diff
@@ -80,6 +84,24 @@ def cli():
+    """
+
+    prompt_kwargs = {
+        "history": FileHistory(os.path.expanduser("~/.prefect/repl-history")),
```
`~/.prefect` should be loaded from the `home_dir` configuration value instead of being hard coded.

madkinsz commented on src/prefect/cli/__init__.py:
```diff
@@ -80,6 +84,24 @@ def cli():
+    Start a REPL
+    """
+    banner = f"""
+Welcome to Prefect REPL
```
```suggestion
Welcome to the Prefect REPL
```

madkinsz commented on src/prefect/cli/__init__.py:
```diff
@@ -80,6 +84,24 @@ def cli():
+@cli.command()
+def repl():
+    """
+    Start a REPL
```
```suggestion
    Start an interactive Prefect CLI session
```

madkinsz commented on requirements.txt:
```diff
@@ -1,4 +1,5 @@
 click >= 7.0
+click-repl == 0.2.0
```
We do not pin dependencies to specific versions, provide a lower bound instead.

madkinsz reviewed (commented):
Hi! Thanks for contributing. I've added some comments that would need to be addressed before we can accept this, but I'm not certain I want to add another dependency for this functionality — once we add it we cannot remove it.

Perhaps instead of adding this directly to our requirements you can try to import the `click_repl` package when the `prefect repl` command is used and if it fails print a message that the additional requirement is needed to use the feature with a `pip install` command?

tirkarthi commented:
Thanks @madkinsz for the review. I have removed click-repl from requirements and added a suggestion in the command while handling `ImportError` as the package is missing. I have also incorporated other suggestions you have added.

madkinsz commented on src/prefect/cli/__init__.py:
```diff
@@ -80,6 +81,38 @@ def cli():
+    except ImportError:
+        print(
+            """
+This command needs click-repl package. Please install it.
+python -m pip install click-repl
```
```suggestion
This command require the 'click-repl' package. Install it first to use this feature:
python -m pip install click-repl
```

madkinsz replied:
Also, we use `pip install ...` without `python -m pip` everywhere else in our documentation, but I don't feel strongly about it here.

tirkarthi replied:
I guess it should be "requires" instead of "require". Please correct me if I am wrong.

madkinsz replied:
👍 looks great. Thanks again!

madkinsz reviewed (commented):
Thanks that looks good. Just a minor rephrasing request on the message.

madkinsz commented:
Did you want to add a changelog entry? You can list this as a `feature` if you'd like

tirkarthi commented:
Thanks, added changelog.
